posthog/models/cohort.py

```diff
@@ -102,29 +102,28 @@ def get_analytics_metadata(self):
             "deleted": self.deleted,
         }
 
-    def calculate_people(self):
+    def calculate_people(self, batch_size=50000, pg_batch_size=5000):
         if self.is_static:
             return
         try:
-            persons_query = self._clickhouse_persons_query()
+            cursor = 0
+            # increase by 1 to test pagination
+            persons_query = self._clickhouse_persons_query(batch_size=batch_size, offset=0)
 
             try:
-                sql, params = persons_query.distinct("pk").only("pk").query.sql_with_params()
+                persons = persons_query.distinct("pk")
             except EmptyResultSet:
-                query = DELETE_QUERY.format(cohort_id=self.pk)
-                params = {}
+                CohortPeople.objects.filter(cohort_id=self.pk).delete()
             else:
-                query = f"""
-                    {DELETE_QUERY};
-                    {UPDATE_QUERY};
-                """.format(
-                    cohort_id=self.pk,
-                    values_query=sql.replace('FROM "posthog_person"', f', {self.pk} FROM "posthog_person"', 1,),
-                )
+                with transaction.atomic():
```

Comment:
Disclaimer: bear with me because I’m not very familiar of the business logic and the full requirements we have.

By only looking at the PR, there are a couple of points I would consider changing in order to reduce reliability and performance risks:

1. we run a `BEGIN` and then we issue a couple of queries to even different datastore systems. One scenario I would consider is `self._clickhouse_persons_query` not returning fast/not returning at all. Here, we would lock the PG table till we hit a timeout and we rollback the full transaction.  

2. as we run this block into a transaction, I’m not actually sure this will solve the overall timeout problem as we are still trying to flush a technically unbounded number of rows in a single commit.

3. batch size: for the ClickHouse SELECT we use 50000 and for the PG INSERT 5000. While I didn’t dig into the schema and checked the avg row size, I suspect 5000 might be still quite a lot of data to insert in one query. 

Let me know what do you think about it, thank you!

Reply:
Agree, 3 is still an issue here.

Stepping back for a sec, my assumed requirements here are:

1. we want to be able to avoid cohorts changing under our feet due to new events coming in, person properties changing etc
2. we want to be able to "refresh" cohorts explicitly
3. whilst we are refreshing a cohort, we want our view of the cohort to be either the stale cohort, or the new cohort, not something in between

(I'm not sure what view of cohort means here, whether that is some list of people we display from postgres, or if it is the view of stats from clickhouse based off this cohort. BTW, how do we calculate these stats on the clickhouse side? Do we use the postgres ids or something?)

It looks like we would, with this change, have a period were we have deleted all cohort people, but are still waiting for the transaction to complete.

Could we instead do without the transaction, but rather version the Cohort/CohortPeople?

This was we can have an atomic switchover without the fuss of worrying putting a transaction around the whole thing. Perhaps having something clean up the db of all the stale cohort people, but this can be handled separately.

We might need some optimistic locking to handle potential concurrency. This way we can also see which cohorts are "dirty", use this to display to the user that it's still waiting to be updated, and if need we can trigger the sync to start such that we are resilient to e.g. pods failing. wdyt?

Reply:
Ooof thought I left this on draft. Thanks for the eyes on this. Agreed with the points above. @hazzadous i looked into versioning briefly as I also thought that would be a cleaner way to do the swapover. However, django's orm was unwieldy so I thought I'd try a different method. Will revisit the versioning approach as the current one would be unreliable

Reply:
🙌 ahaha sorry for the storm of comments so 🤣 @EDsCODE !

Anyway, I don't know what are the business requirements for this but if we are ok to have a slightly inconsistent view of data (that we anyway also have between cohort recalculations executions) something like this could work as well while addressing the issues above:

* fetch all the person_id from ClickHouse (paginated)
* fetch all the person_id from PostgreSQL (paginated)
* diff the two set (in app)
* add / remove person_id from PostgreSQL (paginated)

Reply:
Hmmm, I think that makes sense except we could make it completely consistent if we have versioning? 

So I'm thinking step 4 would insert the cohort-person relationship with an incremented version and once that process is done, we bump the Cohort version +1 to match the new people. Then we can background delete the old versions cohortpeople (don't even have to necessarily). As long as any cohort retrievals made with cohortpeople join on both cohort_id and version, it should be immediately up to date. 

The related issue would be handling the version numbers when multiple cohort recalculations are triggered and maintaining the right version numbers. I think this is what @hazzadous was referring to

Reply:
Yeah either a sequence number, or a string hash of relevant details that signal an update is required. Just something we can use to select from cohortpeople where on. The update process seems pretty simple in this case with no need to do a comparison on what is where.

Apologies for jumping the gun 🔫

Reply:
Even better! 🎉 

> Then we can background delete the old versions cohortpeople (don't even have to necessarily)
Unnecessary for the app behavior but less data we store, the better (as there's an implication in cost, performance, backup/restore time, etc..)

Awesome 💬 folks!